[Static Analyzer CI] Support reporting static analyzer results to results.webkit.org

briannafan opened the pull request:
#### f167dc39b8298efca1aa742a21c7fc976ebd9024
<pre>
[Static Analyzer CI] Support reporting static analyzer results to results.webkit.org
<a href="https://bugs.webkit.org/show_bug.cgi?id=281925">https://bugs.webkit.org/show_bug.cgi?id=281925</a>
<a href="https://rdar.apple.com/138433382">rdar://138433382</a>

Reviewed by Jonathan Bedard.

Take parsed results and format them for the resultsdb API to upload.
Config and build details are provided from buildbot as arguments.
Then, use webkitpy.results.upload to results.webkit.org.

* Tools/Scripts/compare-static-analysis-results:
(parser): Add new options.
(find_diff): Fix empty file logic and return list of all files with issues.
(compare_project_results_to_expectations): Assign an actual and expected result to each file.
(upload_results): Create the results object for upload. Check repo for the current commit. Upload results.
(main): Call upload_results if the upload arguments are passed.

* Tools/Scripts/webkitpy/results/options.py: Support both optparse and argparse!
(add_optargs):
(upload_options):
(upload_args):

Canonical link: <a href="https://commits.webkit.org/285799@main">https://commits.webkit.org/285799@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d919f6f68a5858afea9dfc997ece8cd50acef515

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73893 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26704 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/78227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25131 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76010 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/62455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1107 "Failed to checkout and rebase branch from PR 35588") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/78227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/16456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76960 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/62455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/63571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/78227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/73401 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/62455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/23464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/62455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/21408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79782 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/1210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/1107 "Failed to checkout and rebase branch from PR 35588") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/1353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/63584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11390 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/1174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/1203 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/1190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/generate-dirty-files:
```diff
@@ -179,6 +199,32 @@ def find_all_results(args):
+
+
+def upload_results(options, results):
+    config_options = eval(options.configuration)
```
This is horrifying and scary, can we please not use `eval`?

Given we're already relying on `webkitpy`, can we not just:

```
from webkitpy.port import configuration_options, platform_options
```

and use the same options as elsewhere?

briannafan replied:
Sorry, `eval` got left over from quick prototyping.
I'm not sure adding config and platform options would help much, given we need to pass in other information that isn't covered anyways. Constructing a full config object on the buildbot side allows us to use properties that are already set.

gsnedders commented on Tools/Scripts/generate-dirty-files:
```diff
@@ -179,6 +199,32 @@ def find_all_results(args):
+        style=config_options.get('style') or None,
+        sdk=config_options.get('sdk') or None,
+    )
+    commit = eval(options.commit)
```
Again, let's not use `eval` and run arbitrary Python code.

gsnedders commented on Tools/Scripts/webkitpy/results/options.py:
```diff
@@ -33,3 +34,17 @@ def upload_options(suite_name_flag='suite'):
+    upload_options.add_argument('--build-number', help='The buildbot build number tests are associated with.')
+    upload_options.add_argument('--buildbot-worker', help='The buildbot worker tests were run on.')
+    upload_options.add_argument('--result-report-flavor', help='Optional flag for categorizing test runs which do not fit into other configuration options.')
+    upload_options.add_argument('--' + suite_name_flag, help='Optional flag for overriding reported suite name.', default=None)
```
I wonder if we can somehow make this shared with `upload_options` above; the arguments to `optparse.make_option` and `argparse.ArgumentParser.add_argument` are basically the same?

gsnedders commented on Tools/Scripts/generate-dirty-files:
```diff
@@ -153,14 +173,14 @@ def find_all_results(args):
             sys.stderr.write(f'{e.output}')
             sys.stderr.write(f'Could not find results for {project}\n')
-            return -1
+            break
```
Do we really want to upload nothing in the case where we fail to get results all?

briannafan replied:
Ah I should probably be using `continue` here.

gsnedders commented on Tools/Scripts/generate-dirty-files:
```diff
@@ -135,14 +154,15 @@ def find_project_results(args, project, file_list, results_data, issue_to_report
     for type, count in bug_counts.items():
         print(f'    {type}: {count}')
-    return results_data
+    return results_data, results_for_upload
```
Is there any advantage to return the same object we got passed as a parameter? And then we totally ignore the return value, seemingly.

briannafan replied:
Weird leftovers from old behavior, probably. Thanks for being a fresh set of eyes!

gsnedders commented on Tools/Scripts/generate-dirty-files:
```diff
@@ -86,9 +98,10 @@ def parse_results_file(args, file_path):
+def find_project_results(args, project, file_list, results_data, issue_to_report, results_for_upload):
     bug_counts = {}
     files_per_project = {key: set() for key in CHECKER_MAP.values()}
+    results_for_upload[project] = {}
```
Instead of doing this, why don't we just have a `project_results_for_upload = {}` and then return that, and then at the callsite set `results_for_upload[project]` equal to what we returned?

briannafan replied:
Yeah, I like this better. Thanks!

gsnedders commented on Tools/Scripts/generate-dirty-files:
```diff
@@ -179,6 +199,32 @@ def find_all_results(args):
+        version_name=config_options.get('version_name') or None,
+        style=config_options.get('style') or None,
+        sdk=config_options.get('sdk') or None,
+    )
```
Especially given we already rely on `webkitpy`, maybe we should just use `Port.configuration_for_upload` again?

briannafan replied:
`Port.configuration_for_upload` relies on having a port, which these tests don't have. I'm open to solutions that don't rely on passing in configs as an option but I don't know if there are any good ones.

gsnedders reviewed (changes requested):
I feel like we may as well bite the bullet and just use `webkitpy.port.Port` here, and then we get a lot of this for free?

JonWBedard commented on Tools/Scripts/generate-dirty-files:
```diff
@@ -121,6 +134,12 @@ def find_project_results(args, project, file_list, results_data, issue_to_report
+            project_results_for_upload[CHECKER_MAP[bug_type]] = {}
+
+        if not file_name in project_results_for_upload[CHECKER_MAP[bug_type]].keys():
+            project_results_for_upload[CHECKER_MAP[bug_type]][file_name] = {'actual': 'FAIL'}
```
This seems like we're going to end up with a trie of `bug-type.file_name = RESULT`. First, it seems like it should be reversed (ie, if I start search for a file, I get all types of failures for that file). Second, it doesn't look like we're uploading successful runs for files. We should. Having successful runs of files is very useful, it lets us quickly eliminate problems like "were we even running this test". Successful runs is the whole reason we use the trie design to being with, it makes uploads much smaller.

briannafan replied:
Per offline conversation:
- We'll upload results by `project.file_name.bug_type = RESULT`
- Due to the difficulty of obtaining a full list of all files that static analysis is run on, we won't be uploading successful runs. However, we are able to upload successes when the expectation is a failure
- Changes also include uploading expected results for every file that is reported in a run!

briannafan committed:
[Static Analyzer CI] Support reporting static analyzer results to results.webkit.org
https://bugs.webkit.org/show_bug.cgi?id=281925
rdar://138433382

Reviewed by Jonathan Bedard.

Take parsed results and format them for the resultsdb API to upload.
Config and build details are provided from buildbot as arguments.
Then, use webkitpy.results.upload to results.webkit.org.

* Tools/Scripts/compare-static-analysis-results:
(parser): Add new options.
(find_diff): Fix empty file logic and return list of all files with issues.
(compare_project_results_to_expectations): Assign an actual and expected result to each file.
(upload_results): Create the results object for upload. Check repo for the current commit. Upload results.
(main): Call upload_results if the upload arguments are passed.

* Tools/Scripts/webkitpy/results/options.py: Support both optparse and argparse!
(add_optargs):
(upload_options):
(upload_args):

Canonical link: https://commits.webkit.org/285799@main
